tests/test_modeling_rembert.py

```diff
@@ -0,0 +1,499 @@
+
+    def test_for_question_answering(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_for_question_answering(*config_and_inputs)
+
+    def test_for_sequence_classification(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_for_sequence_classification(*config_and_inputs)
+
+    def test_for_token_classification(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_for_token_classification(*config_and_inputs)
+
+    def test_model_as_decoder(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs_for_decoder()
+        self.model_tester.create_and_check_model_as_decoder(*config_and_inputs)
+
+    def test_model_as_decoder_with_default_input_mask(self):
+        # This regression test was failing with PyTorch < 1.3
+        (
+            config,
+            input_ids,
+            token_type_ids,
+            input_mask,
+            sequence_labels,
+            token_labels,
+            choice_labels,
+            encoder_hidden_states,
+            encoder_attention_mask,
+        ) = self.model_tester.prepare_config_and_inputs_for_decoder()
+
+        input_mask = None
+
+        self.model_tester.create_and_check_model_as_decoder(
+            config,
+            input_ids,
+            token_type_ids,
+            input_mask,
+            sequence_labels,
+            token_labels,
+            choice_labels,
+            encoder_hidden_states,
+            encoder_attention_mask,
+        )
+
+    @slow
+    def test_model_from_pretrained(self):
+        for model_name in REMBERT_PRETRAINED_MODEL_ARCHIVE_LIST[:1]:
+            model = RemBertModel.from_pretrained(model_name)
+            self.assertIsNotNone(model)
+
+
+@require_torch
+class RemBertModelIntegrationTest(unittest.TestCase):
+    @slow
+    def test_inference_model(self):
+        # Test exact values at the last hidden layer
+        model = RemBertModel.from_pretrained("google/rembert")
+        input_ids = torch.tensor([[312, 56498, 313, 2125, 313]])
+        segment_ids = torch.tensor([[0, 0, 0, 1, 1]])
+        output = model(input_ids, token_type_ids=segment_ids, output_hidden_states=True)
+
+        hidden_size = 1152
+
+        expected_shape = torch.Size((1, 5, hidden_size))
+        self.assertEqual(output["last_hidden_state"].shape, expected_shape)
+
+        expected_implementation = torch.tensor(
+            [
+                [
+                    [0.0754, -0.2022, 0.1904],
+                    [-0.3354, -0.3692, -0.4791],
+                    [-0.2314, -0.6729, -0.0749],
+                    [-0.0396, -0.3105, -0.4234],
+                    [-0.1571, -0.0525, 0.5353],
+                ]
+            ]
+        )
+
+        # Running on the original tf implementation gives slightly different results here.
```

Comment:
Hmm, it would be nice to have an integration test with 1e-3 tolerance only. IMO, it would be important to find out what the difference is before merging

Reply:
Yes, I agree and have tried a few things to bridge the gap, including running things layer by layer.

We know that the discrepancy starts from the first layer onwards. Results are equal for the tf and hf version after we upproject the embeddings, but differ after the first layer.

Unfortunately, I no longer have access to the tf codebase inside google and it has been hard to get sublayer outputs to understand which op exactly explains the difference. I also tried to run a home-baked version of the tf codebase but had trouble installing a old enough version of tf + running it. Ultimately, I mostly looked at the fine-tuning results, which are good, though there is still a gap there. @hwchung might be able to help but he has been pretty busy lately.